core/log/profiler_hook_summary.cpp

```diff
@@ -0,0 +1,406 @@
+        const auto elapsed_ns =
+            std::chrono::duration_cast<std::chrono::nanoseconds, int64>(
+                now - partial_entry.second)
+                .count();
+        entry.count++;
+        entry.inclusive_ns += elapsed_ns;
+        entry.exclusive_ns += elapsed_ns;
+        if (!stack.empty()) {
+            entries[stack.back().first].exclusive_ns -= elapsed_ns;
+        }
+        overhead_ns +=
+            std::chrono::duration_cast<std::chrono::nanoseconds, int64>(
+                clock::now() - now)
+                .count();
+    }
+
+    const std::string& get_top_name() const
+    {
+        return entries[stack.back().first].name;
+    }
+};
+
+
+struct nested_summary {
+    struct entry {
+        int64 name_id;
+        int64 node_id;
+        int64 parent_id;
+        int64 elapsed_ns{};
+        int64 count{};
+
+        entry(int64 name_id, int64 node_id, int64 parent_id)
+            : name_id{name_id}, node_id{node_id}, parent_id{parent_id}
+        {}
+    };
+
+    struct partial_entry {
+        int64 name_id;
+        int64 node_id;
+        time_point start;
+
+        partial_entry(int64 name_id, int64 node_id, time_point start)
+            : name_id{name_id}, node_id{node_id}, start{start}
+        {}
+    };
+
+    struct pair_hash {
+        int64 operator()(std::pair<int64, int64> pair) const
+        {
+            return pair.first ^ (pair.second << 32);
+        }
+    };
+
+    int64 overhead_ns{};
+    std::vector<partial_entry> stack;
+    std::unordered_map<std::pair<int64, int64>, int64, pair_hash> node_map;
+    std::unordered_map<std::string, int64> name_map;
+    std::vector<entry> nodes;
+    std::vector<std::string> names;
+    bool broken{};
+    bool check_nesting{};
+    std::mutex mutex{};
+
+    nested_summary() { push("total"); }
+
+    int64 get_or_add_name_id(const char* name)
+    {
+        const auto it = name_map.find(name);
+        if (it != name_map.end()) {
+            return it->second;
+        }
+        const auto name_id = static_cast<int64>(names.size());
+        name_map.emplace_hint(it, name, name_id);
+        names.push_back(name);
+        return name_id;
+    }
+
+    int64 get_or_add_node_id(int64 name_id)
+    {
+        const auto parent_id = get_parent_id();
+        const auto pair = std::make_pair(name_id, parent_id);
+        const auto it = node_map.find(pair);
+        if (it != node_map.end()) {
+            return it->second;
+        }
+        const auto node_id = static_cast<int64>(nodes.size());
+        node_map.emplace_hint(it, pair, node_id);
+        nodes.emplace_back(name_id, node_id, parent_id);
+        return node_id;
+    }
+
+    int64 get_parent_id() const
+    {
+        return stack.empty() ? int64{-1} : stack.back().node_id;
+    }
+
+    void push(const char* name)
+    {
+        if (broken) {
+            return;
+        }
+        const auto now = clock::now();
+        std::lock_guard<std::mutex> guard{mutex};
```

Comment:
yes, otherwise we might be concurrently reallocating the vector or writing to the hashmap, which is a typical race condition. This can only occur in the OpenMP executor or user code, but better be safe than sorry ;)